dspace-api/src/test/java/org/dspace/curate/XmlWorkflowCuratorServiceImplTest.java

```diff
@@ -0,0 +1,277 @@
+import org.dspace.xmlworkflow.storedcomponents.XmlWorkflowItem;
+import org.dspace.xmlworkflow.storedcomponents.service.ClaimedTaskService;
+import org.junit.BeforeClass;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
+import org.springframework.context.ApplicationContext;
+
+/**
+ *
+ * @author mwood
+ */
+public class XmlWorkflowCuratorServiceImplTest
+        extends AbstractUnitTest {
+    private static final String GROUP_ONE = "one";
+    private static final String TASK_NOOP = "noop";
+    private static final String NULL_WORKFLOW = "nullWorkflow";
+
+    // Configure the workflow task mapping.
+    private static final String TASK_CONFIG_TEMPLATE
+            = ("<?xml version='1.0' encoding='UTF-8'?>"
+            + "<workflow-curation>"
+            + "  <taskset-map>"
+            + "    <mapping collection-handle='default' taskset='default'/>"
+            + "  </taskset-map>"
+            + "  <tasksets>"
+            + "    <taskset name='default'>"
+            + "      <flowstep name='%s'>"
+            + "        <task name='%s'/>"
+            + "      </flowstep>"
+            + "    </taskset>"
+            + "  </tasksets>"
+            + "</workflow-curation>");
+
+    private static CollectionService collectionService;
+
+    private static ConfigurationService dspaceConfiguration;
+
+    private static EPersonService epersonService;
+
+    private static GroupService groupService;
+
+    private static XmlWorkflowServiceFactory wsf;
+
+    private static XmlWorkflowService workflowService;
+
+    private static AutowireCapableBeanFactory beanFactory;
+
+    @BeforeClass
+    public static void setUpClass()
+            throws SQLException {
+        // Get some necessary service instances.
+        dspaceConfiguration = DSpaceServicesFactory.getInstance().getConfigurationService();
+        collectionService = ContentServiceFactory.getInstance().getCollectionService();
+        EPersonServiceFactory esf = EPersonServiceFactory.getInstance();
+        epersonService = esf.getEPersonService();
+        groupService = esf.getGroupService();
+        wsf = XmlWorkflowServiceFactory.getInstance();
+        workflowService = wsf.getXmlWorkflowService();
+
+        beanFactory = kernelImpl.getServiceManager()
+                .getServiceByName(ApplicationContext.class.getName(), ApplicationContext.class)
+                .getAutowireCapableBeanFactory();
+
+        // Ensure that a known task exists with a known name.
+        final String CURATION_TASK_MAP_NAME
+                = "plugin.named.org.dspace.curate.CurationTask";
+        dspaceConfiguration.setProperty(CURATION_TASK_MAP_NAME, null);
+        dspaceConfiguration.addPropertyValue(CURATION_TASK_MAP_NAME,
+                NoOpCurationTask.class.getCanonicalName() + " = " + TASK_NOOP);
+
+        // XXX Try to make Builders work in unit test.
+        AbstractBuilder.init();
```

Comment:
@mwoodiupui : Just to second this thought (copying what I mentioned in Slack)... as has been discussed in #3055, we are trying to move away from Unit Tests creating any content.  So, the fact that this test class extends `AbstractUnitTests` _and uses Builders_ is a bit of an anti-pattern.  (It's something we are trying to move away from.)

So, there are a few options here... either we look at refactoring this test to no longer rely on creating content (if that's even possible), or we could look at simply changing this to an IT and extending `AbstractIntegrationTestWithDatabase` (which is where Builders are welcome).  Or, we could just keep this as is for now & clean it up as part of #3055 if neither of those options are easy right now.

Reply:
A better test would indeed be a welcome development.  You're seeing some of the rubble left after two weeks of bafflement, frustration and outrage as design after design failed, telling me that objects I had just created did not exist.  The current test code is in the "at last, it works! don't touch anything!" state.  The interior of the configurable workflow system is deeply mysterious, at least to me, and I hadn't found any enlightening examples of working with it.  I suppose this means that I should go tramping through the entire workflow subsystem, code machete in hand, and write a lot of documentation when the design becomes clear to me.

I'm happy to hear that there is movement to better organize our test practices.  This class was changed from IT to UT late in the game, when I felt that the IT framework was fighting me and I needed to simplify.  I stuck with the Builders because they presumably know the proper sequence of actions to set up the necessary objects, and it was becoming clear that I was missing some of that.  Possibly for UTs we need a parallel set of Maker classes to build in-memory objects *not* backed by the database, if we are going to continue preventing arbitrary code from instantiating model objects.  We also should do some refactoring to make some tested classes more testable.

Reply:
@mwoodiupui : Would you be willing to try changing this to extend `AbstractIntegrationTestWithDatabase` one last time?  I'm very curious why this would _work_ as an `AbstractUnitTest` with Builders, but _not work_ as an `AbstractIntegrationTestWithDatabase` with Builders.  That seems unexpected.

Reply:
I've scrapped that whole test suite and written a new IT class, extending `AbstractIntegrationTestWith Database`.  I was helped by having an example to look at.

Unfortunately I forgot why I was writing the test in the REST layer, and when the resulting code didn't seem to have any necessary connection with REST I factored it out and moved it back to dspace-api.

In the course of refactoring I discovered that (a) I needed two Spring initializers, which I copied from `dspace-server-webapp` to `dspace-api`, and (b) those initializers, while they say "Spring Boot" all over them, really don't depend on Boot at all.  I have wondered whether they really ought to be part of the scaffolding in `dspace-services`.  At any rate there's no reason why they can't be available to our whole constellation of projects, and having JUnit run them makes the test code quite a bit cleaner than trying to inject dependencies manually.